geps/gep-1731/index.md

```diff
@@ -0,0 +1,409 @@
+
+* `all-retryable-errors` Retry request for any error that are considered
+retryable. This currently activates "conn-failure", "empty-response", "junk-response", "response-timeout", "0rtt-rejected", "500", "502", "503", and "504".
+
+The [`option redispatch`](https://docs.haproxy.org/3.0/configuration.html#4.2-option%20redispatch) configuration can be used to distribute retries across multiple backend servers, allowing the proxy to break cookie or consistent hash based persistence and redistribute them to a working server.
+
+#### Traefik
+
+Offers a [Retry middleware](https://doc.traefik.io/traefik/middlewares/http/retry/) allowing configuration of number of attempts and initial interval.
+
+Reissues requests a given number of times to a backend server if that server does not reply. As soon as the server answers, the middleware stops retrying, _regardless of the response status_. Has an optional configuration to enable an exponential backoff.
+
+* `attempts` Defines how many times the request should be retried.
+
+* `initialInterval` Defines the first wait time in the backoff series. The maximum interval is calculated as twice the `initialInterval`. If unspecified, requests will be retried immediately. The value should be provided in seconds or as a valid duration format.
+
+Supports configuration of a [Circuit Breaker](https://doc.traefik.io/traefik/middlewares/http/circuitbreaker/) which could possibly be used to implement budgeted retries. Each router gets its own instance of a given circuit breaker. One circuit breaker instance can be open while the other remains closed: their state is not shared. This is the expected behavior, we want you to be able to define what makes a service healthy without having to declare a circuit breaker for each route.
+
+#### linkerd2-proxy
+
+Linkerd supports a configurable [retry budget](https://linkerd.io/2.15/features/retries-and-timeouts/) per-route using its [ServiceProfile](https://linkerd.io/2.15/reference/service-profiles/) CRD. The default budget is 20%, but Linkerd also supports a certain number of "extra" retries every second (10 by default), for a better user experience with low-traffic services.
+
+Linkerd is unique in that it does not currently support counted retries, although this is an area of active development.
+
+#### F5 BIG-IP
+
+TODO
+
+#### Pipy
+
+TODO
+
+#### Litespeed WebADC
+
+TODO
+
+## API
+
+### Go
+
+```golang
+type HTTPRouteRule struct {
+    // Retry defines the configuration for when to retry an HTTP request.
+    //
+    // Support: Extended
+    //
+    // +optional
+    // <gateway:experimental>
+    Retry *HTTPRouteRetry `json:"retry,omitempty"`
+
+    // ...
+}
+
+// HTTPRouteRetry defines retry configuration for an HTTPRoute.
+//
+// Implementations MUST retry on connection errors (disconnect, reset, timeout,
+// TCP failure) if a retry stanza is configured.
```

Comment:
Feedback wanted on this

Reply:
Specifically, NGINX and HAProxy split out `timeout` from other connection errors and discourage retrying on timeouts, but I believe this advice is less applicable on modern Kubernetes deployments, see Alternatives section at the end for more detail.

Reply:
I'm probably missing something obvious - why wouldn't this class of retry reasons just be populated in `codes`?

Reply:
Because these aren't status codes, and the precise semantics vary between implementations - for example, Envoy groups `timeout` into the same bucket as `disconnect` and `reset`.

Reply:
☝️ Gotta agree with Mike here. Meshes make this even weirder: with Linkerd, for example, the state of the connection from one proxy to another is managed independently of requests arriving from the application workload.

Reply:
To clarify, I agree that there are distinct reasons to retry that aren't just HTTP status codes, my concern was more that I implemented this statement as a blanket requirement that all implementations retry on these events if _any_ part of the retry configuration is present. Let's say I want to have code that _only_ retries if I get a specific response, say 502. As written, this seems to prevent that from being expressible. 

As far as I can tell, every proxy makes this behavior optional, so I expected this to be expressed by some kind of entry in the `codes` list, such as `gateway.envoyproxy.io/disconnect` or similar. If we could find some kind of portable way to express this across dataplanes, that would be even better, I'm just a bit uncomfortable with building this behavior in as a hidden (and likely surprising) default.

Reply:
> I interpreted this statement as a blanket requirement that all implementations retry on these events if any part of the retry configuration is present. Let's say I want to have code that only retries if I get a specific response, say 502. As written, this seems to prevent that from being expressible.

Yes, that's a correct interpretation, with a few motivating factors behind this:
- The capability of splitting up these failure modes granularly varies significantly between implementations.
  - Envoy includes `TCP connection failure` as `connect-failure` in the `5xx` shorthand, but groups `disconnect`, `reset` and `timeout` separately under `reset`, I'm unclear if/how these are distinct.
  - NGINX defines `error` as `An error occurred while establishing a connection with the server, passing a request to it, or reading the response header.` and `timeout` separately as `A timeout has occurred while establishing a connection with the server, passing a request to it, or reading the response header.`
  - HAProxy defines `conn-failure` as `Retry when the connection or the SSL handshake failed and the request could not be sent.` setting this as default behavior, and `response-timeout` separately with a warning it may be inadvisable to retry (see note near bottom of GEP, my theory is that this warning is more applicable to legacy deployment models with a small number of static servers).
- Making this a default behavior reduces the boilerplate required for typical usage.
- I haven't seen _any_ documentation or user story indicating that _not_ retrying on TCP connection failures is a desirable property.

I'd be open to adding something like a separate `excludeRetryOnTimeout` field (open to suggestions for better name) that NGINX and HAProxy (but not Envoy) should be able to implement.

Reply:
Overall I'm uncomfortable with adding default/implied behavior here. If I add configuration that I want to retry exclusively on 502s, I would be surprised if retries happened on other cases. 

I definitely get the idea that _most_ use cases would benefit from this class of retries, but I'm not sure that justifies defaulting to behavior that I think could be surprising to users. Maybe an alternative would be to bucket this class of timeouts into a well known code, and maybe even add that as a default if nothing else is specified in the list? (The defaulting is still a bit messy though as it would make it particularly difficult to say "never retry").

Reply:
> (The defaulting is still a bit messy though as it would make it particularly difficult to say "never retry").

Hmm, I had considered "never retry" as not likely portable - it appeared only HAProxy supported this explicitly, I suppose setting `attempts: 0` could be a reasonable way to do that though?

> Envoy will default to retrying one time unless explicitly specified
https://www.envoyproxy.io/docs/envoy/latest/configuration/http/http_filters/router_filter#x-envoy-max-retries

> I definitely get the idea that most use cases would benefit from this class of retries, but I'm not sure that justifies defaulting to behavior that I think could be surprising to users.

I think behavior that would be beneficial to most use cases is a prime candidate to be a default - Envoy implicitly lumps connection failures in with `5xx` which I think is even more confusing. I could see a case for adding an opt-out like `excludeRetryOnConnectionFailure` as distinct from the proposed opt-out for retry on timeouts I suggested in 0e533c6e8117561535f5be5ebfc39023f6dc1178, but I think only NGINX (and maybe HAProxy?) would be able to implement that.

Reply:
I don't see a practical way to portably talk about retries based on connection state, _especially_ as we move further and further away from HTTP/1 into a world where multiplexing unrelated requests over a long-lived connection is the norm... or a world in which HTTP/3 becomes more common in the cluster.

Maybe just add language to the tune of

```
Implementations MAY additionally retry for reasons related to connection state changes (e.g. a connection timing out or breaking); such behavior is implementation-specific by definition. Applications SHOULD NOT be designed to rely on such behavior as it may be changed by future work in this area.
```

???

Reply:
The problem I have with MAY is that it makes this behavior entirely unpredictable (SHOULD feels okay and maybe we add an additional separate conformance flag), plus effectively makes writing a conformance test supportable by Traefik impossible even though it supports a functional subset of this feature.

Reply:
I think my preferred path would be to say that implementation-specific defaults for these categories may vary (similar to timeouts), but to leave open a path to specifying them in one large bucket defined by Gateway API. I think grouping all of these into a single bucket into the future could help with the overall UX while also avoiding surprises. The biggest thing I'm trying to avoid is _requiring_ default behavior that at least some users will not find desirable.

I do think that writing conformance tests is going to be tricky here. I don't want to block this entire GEP on this relatively small detail though, as I think we can refine how this part of the spec is worded in the next phase.

Reply:
@mikemorris do you mind adding some kind of `UNRESOLVED` block around this specific part of the spec? So that we don't miss this in the next phase?

Reply:
Switched recommendation to SHOULD, added conformance flags and warning in 510b708b8e793f98bd2704b8399fd8bcb2f86ae5